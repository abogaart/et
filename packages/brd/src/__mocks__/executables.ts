
import { ExecutableLookupService } from '@bloomreach/cli-services';

export function mockExecutable(name: string, path: string, version: string): any {
  return {
    getName: jest.fn().mockReturnValue(name),
    getPath: jest.fn().mockReturnValue(path),
    getVersion: jest.fn().mockReturnValue(version),
  };
}

export const mockExecutableLookupService: jest.Mock & ExecutableLookupService = {
  create: jest.fn()
    .mockImplementation((name: string, path: string) => mockExecutable(name, path, `${name}-custom-path-version`)),
  lookup: jest.fn(),
};

export const mockCliServices: jest.Mock = {
  get executableLookupService(): ExecutableLookupService {
    return mockExecutableLookupService;
  },
};

export function mockLookupExecutable(...execs): void {
  mockExecutableLookupService.lookup.mockImplementation((name: string) => {
    const r = execs.find((exec) => name === exec.name);
    return r ? r.execs : [];
  });
}
