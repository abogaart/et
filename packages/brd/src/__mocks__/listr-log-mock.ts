export function mockListerLog() {
  jest.mock('listr-verbose-renderer');
  // listr-verbose-renderer is a transitive dependency of listr and is installed in the node_modules of the listr module
  // If we make it an implicit dependency, we end up forking the wrong class (from <project-dir>/node_modules instead of <project-dir>/node_modules/listr/node_modules)
  // tslint:disable-next-line:no-implicit-dependencies
  const { log } = require('listr-verbose-renderer/lib/utils');
  log.mockImplementation((_cfg: any, message: 'string') => {
    process.stdout.write(message + '\n');
  });

  log;
}
