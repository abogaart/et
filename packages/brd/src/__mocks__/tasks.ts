import { Tasks } from '@bloomreach/cli-tasks';

export function mockChoices(index: number): void {
  Tasks.choice = jest.fn().mockImplementation((
    _message: string,
    choices: { name: string; value: any }[],
    next: (choice: any) => Promise<void>,
  ) => next(choices[index].value));
}

export function mockCustomPath(...ids: string[]): void {
  Tasks.customPath = jest.fn().mockImplementation((
    confirmMsg: string,
    _pathLabel: string,
    next: (path: string) => Promise<void>,
  ) => {
    const pathId = ids.find((id: string) => confirmMsg.includes(id));
    return pathId
      ? next(`/custom-path/${pathId}`)
      : next('/custom-path');
  });
}
