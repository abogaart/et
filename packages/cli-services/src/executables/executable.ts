export interface Executable {
  getName(): string;
  getPath(): string;
  getVersion(): Promise<string>;
}

export abstract class Executable implements Executable {
  private version: string | undefined;

  constructor(private name: string, private path: string) {
  }

  getName(): string {
    return this.name;
  }

  getPath(): string {
    return this.path;
  }

  async getVersion(): Promise<string> {
    if (this.version === undefined) {
      try {
        this.version = await this.parseVersion(this.path);
      } catch (ignore) {
        this.version = '';
      }
    }
    return this.version;
  }

  protected abstract async parseVersion(path: string): Promise<string>;
}
