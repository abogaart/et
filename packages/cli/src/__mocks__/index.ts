import * as path from 'path';

import { EtCommand, EtFlags, EtContext } from '../base';

export class TestCommand<Flags extends EtFlags> extends EtCommand<Flags> {
  static flags = EtCommand.flags;

  run = jest.fn();

  get context(): EtContext<Flags> {
    return this.getContext();
  }
}

export function fixturesPath(...paths): string {
  return path.resolve(__dirname, '../__fixtures__', ...paths);
}
