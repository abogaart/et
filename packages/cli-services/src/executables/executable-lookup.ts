import { findLinkedFile, tryWhich } from '@bloomreach/cli-utils';
import { Executable } from './executable';

export interface ExecutableLookup {
  resolve(): Promise<Executable[]>;
}

export abstract class ExecutableLookup implements ExecutableLookup {
  constructor(protected name: string) {}

  async resolve(): Promise<Executable[]> {
    const execs = [];

    const fromWhich = await tryWhich(this.name);
    if (fromWhich !== null) {
      execs.push(this.createExecutable(this.name));
      execs.push(this.createExecutable(fromWhich));
      execs.push(this.createExecutable(findLinkedFile(fromWhich)));
    }

    // TODO: filter dups
    return execs;
  }

  abstract createExecutable(path: string): Executable;
}
