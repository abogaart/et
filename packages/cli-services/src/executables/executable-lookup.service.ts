import { ExecutableLookup } from './executable-lookup';
import { Executable } from './executable';

export interface ExecutableLookupService {
  create(exec: string, path: string): Executable;
  lookup(exec: string): Promise<Executable[]>;
  register(exec: string, lookup: ExecutableLookup): void;
}

export class ExecutableLookupService implements ExecutableLookupService {
  private lookups = new Map<string, ExecutableLookup>();

  create(exec: string, path: string): Executable {
    const lookup = this.lookups.get(exec);
    if (!lookup) {
      throw new Error(`Failed to create executable ${exec} with path ${path}. No executable lookup defined for ${exec}.`);
    }
    return lookup.createExecutable(path);
  }

  async lookup(exec: string): Promise<Executable[]> {
    const lookup = this.lookups.get(exec);
    if (!lookup) {
      return Promise.resolve([]);
    }
    return lookup.resolve();
  }

  register(exec: string, lookup: ExecutableLookup): void {
    this.lookups.set(exec, lookup);
  }
}
