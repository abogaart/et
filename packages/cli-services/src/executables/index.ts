import { ExecutableLookupService } from './executable-lookup.service';
import { GitExecutable } from './git-executable';
import { JavaExecutable } from './java-executable';
import { MavenExecutable } from './maven-executable';
import { GitExecutableLookup } from './git-executable-lookup';
import { JavaExecutableLookup } from './java-executable-lookup';
import { MavenExecutableLookup } from './maven-executable-lookup';

export { ExecutableLookupService } from './executable-lookup.service';
export { Executable } from './executable';
export { ExecutableLookup } from './executable-lookup';
export { GitExecutable } from './git-executable';
export { GitExecutableLookup } from './git-executable-lookup';
export { JavaExecutable } from './java-executable';
export { JavaExecutableLookup } from './java-executable-lookup';
export { MavenExecutable } from './maven-executable';
export { MavenExecutableLookup } from './maven-executable-lookup';

export const executableLookupService = new ExecutableLookupService();
executableLookupService.register(GitExecutable.CMD, new GitExecutableLookup());
executableLookupService.register(JavaExecutable.CMD, new JavaExecutableLookup());
executableLookupService.register(MavenExecutable.CMD, new MavenExecutableLookup());
