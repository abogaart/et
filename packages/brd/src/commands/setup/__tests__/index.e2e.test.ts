import { test } from '@oclif/test';
import * as cliUtilsModule from '@bloomreach/cli-utils';
import { jestExpect } from '@bloomreach/test-utils';
import { Tasks } from '@bloomreach/cli-tasks';

import { mockListerLog } from '../../../__mocks__/listr-log-mock';
import { mockLookupExecutable, mockExecutable } from '../../../__mocks__/executables';
import { mockChoices, mockCustomPath } from '../../../__mocks__/tasks';

// setup mocks
jest.mock('listr-verbose-renderer/lib/utils');
mockListerLog();

// eslint-disable-next-line global-require
jest.mock('@bloomreach/cli-services', () => require('../../../__mocks__/executables').mockCliServices);

Tasks.choice = jest.fn();
Tasks.fail = jest.fn();

jest.spyOn(cliUtilsModule, 'writeJson').mockReturnValue(true);

const gitExec = {
  name: 'git',
  execs: [
    mockExecutable('git', 'git', 'git-global-version'),
    mockExecutable('git', '/abs/path/to/git', 'git-abs-version'),
    mockExecutable('git', '/symlink/path/to/git', 'git-symlink-version'),
  ],
};

const javaExec = {
  name: 'java',
  execs: [
    mockExecutable('java', 'java', 'java-global-version'),
    mockExecutable('java', '/abs/path/to/java', 'java-abs-version'),
    mockExecutable('java', '/symlink/path/to/java', 'java-symlink-version'),
  ],
};

const mvnExec = {
  name: 'mvn',
  execs: [
    mockExecutable('mvn', 'mvn', 'maven-global-version'),
    mockExecutable('mvn', '/abs/path/to/mvn', 'maven-abs-version'),
    mockExecutable('mvn', '/symlink/path/to/mvn', 'maven-symlink-version'),
  ],
};

const cmd = ['setup', '--config', 'test-config'];

describe('setup', () => {
  beforeEach(() => {
    Tasks.choice.mockClear();
    Tasks.fail.mockClear();

    mockLookupExecutable(gitExec, javaExec, mvnExec);
  });

  describe('initial', () => {
    test
      .stdout()
      .command([...cmd, '--default'])
      .it('picks first path of git, java and mvn executables in default mode', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
        jestExpect(Tasks.choice).not.toHaveBeenCalled();
      });

    test
      .do(() => mockChoices(0))
      .stdout()
      .command(cmd)
      .it('finds global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
        jestExpect(Tasks.choice).toHaveBeenCalledTimes(3);
      });

    test
      .do(() => mockChoices(1))
      .stdout()
      .command(cmd)
      .it('can select absolute path from global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
        jestExpect(Tasks.choice).toHaveBeenCalledTimes(3);
      });

    test
      .do(() => mockChoices(2))
      .stdout()
      .command(cmd)
      .it('can select symlink resolved absolute path from global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
        jestExpect(Tasks.choice).toHaveBeenCalledTimes(3);
      });

    test
      .do(() => {
        mockLookupExecutable(
          { ...gitExec, execs: [] },
          { ...javaExec, execs: [] },
          { ...mvnExec, execs: [] },
        );
        mockCustomPath('git', 'java', 'mvn');
      })
      .stdout()
      .command(cmd)
      .it('present option to set custom paths', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
      });

    test
      .do(() => {
        const invalidVersion = '';
        mockLookupExecutable(
          { name: 'git', execs: [mockExecutable('git', '/invalid/git/path', invalidVersion)] },
          { name: 'java', execs: [mockExecutable('java', '/invalid/java/path', invalidVersion)] },
          { name: 'mvn', execs: [mockExecutable('mvn', '/invalid/maven/path', invalidVersion)] },
        );
      })
      .stdout()
      .stderr()
      .command(cmd)
      .it('prints an error if custom path is invalid', () => {
        jestExpect(Tasks.fail.mock.calls).toHaveLength(3);
        jestExpect(Tasks.fail.mock.calls[0][0]).toMatchSnapshot();
        jestExpect(Tasks.fail.mock.calls[1][0]).toMatchSnapshot();
        jestExpect(Tasks.fail.mock.calls[2][0]).toMatchSnapshot();
      });
  });

  describe('already installed', () => {
    test
      .stdout()
      .command(['setup', '--config', 'src/commands/setup/__tests__/__fixtures__/et-installed'])
      .it('print merged config if already stalled', (ctx: any) => {
        jestExpect(ctx.stdout).toMatchSnapshot();
      });
  });
});
