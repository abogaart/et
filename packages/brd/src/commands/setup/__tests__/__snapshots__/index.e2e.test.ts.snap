// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup already installed print merged config if already stalled 1`] = `
"The \\"brd\\" command is installed with the following properties
{
  \\"installed\\": true,
  \\"version\\": \\"bet-installed\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;

exports[`setup initial can select absolute path from global installed git, java and mvn executables 1`] = `
"
The \\"brd\\" command has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
git found at '/abs/path/to/git' (vgit-abs-version) [title changed]
git found at '/abs/path/to/git' (vgit-abs-version) [completed]
Detect java executable [started]
java found at '/abs/path/to/java' (vjava-abs-version) [title changed]
java found at '/abs/path/to/java' (vjava-abs-version) [completed]
Detect mvn executable [started]
mvn found at '/abs/path/to/mvn' (vmaven-abs-version) [title changed]
mvn found at '/abs/path/to/mvn' (vmaven-abs-version) [completed]

Setup has successfully finished
{
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"/abs/path/to/git\\"
  },
  \\"java\\": {
    \\"path\\": \\"/abs/path/to/java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"/abs/path/to/mvn\\"
  }
}
"
`;

exports[`setup initial can select symlink resolved absolute path from global installed git, java and mvn executables 1`] = `
"
The \\"brd\\" command has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
git found at '/symlink/path/to/git' (vgit-symlink-version) [title changed]
git found at '/symlink/path/to/git' (vgit-symlink-version) [completed]
Detect java executable [started]
java found at '/symlink/path/to/java' (vjava-symlink-version) [title changed]
java found at '/symlink/path/to/java' (vjava-symlink-version) [completed]
Detect mvn executable [started]
mvn found at '/symlink/path/to/mvn' (vmaven-symlink-version) [title changed]
mvn found at '/symlink/path/to/mvn' (vmaven-symlink-version) [completed]

Setup has successfully finished
{
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"/symlink/path/to/git\\"
  },
  \\"java\\": {
    \\"path\\": \\"/symlink/path/to/java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"/symlink/path/to/mvn\\"
  }
}
"
`;

exports[`setup initial finds global installed git, java and mvn executables 1`] = `
"
The \\"brd\\" command has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
git found at 'git' (vgit-global-version) [title changed]
git found at 'git' (vgit-global-version) [completed]
Detect java executable [started]
java found at 'java' (vjava-global-version) [title changed]
java found at 'java' (vjava-global-version) [completed]
Detect mvn executable [started]
mvn found at 'mvn' (vmaven-global-version) [title changed]
mvn found at 'mvn' (vmaven-global-version) [completed]

Setup has successfully finished
{
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;

exports[`setup initial picks first path of git, java and mvn executables in default mode 1`] = `
"
The \\"brd\\" command has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
git found at 'git' (vgit-global-version) [title changed]
git found at 'git' (vgit-global-version) [completed]
Detect java executable [started]
java found at 'java' (vjava-global-version) [title changed]
java found at 'java' (vjava-global-version) [completed]
Detect mvn executable [started]
mvn found at 'mvn' (vmaven-global-version) [title changed]
mvn found at 'mvn' (vmaven-global-version) [completed]

Setup has successfully finished
{
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;

exports[`setup initial present option to set custom paths 1`] = `
"
The \\"brd\\" command has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
git found at '/custom-path/git' (vgit-custom-path-version) [title changed]
git found at '/custom-path/git' (vgit-custom-path-version) [completed]
Detect java executable [started]
java found at '/custom-path/java' (vjava-custom-path-version) [title changed]
java found at '/custom-path/java' (vjava-custom-path-version) [completed]
Detect mvn executable [started]
mvn found at '/custom-path/mvn' (vmvn-custom-path-version) [title changed]
mvn found at '/custom-path/mvn' (vmvn-custom-path-version) [completed]

Setup has successfully finished
{
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"/custom-path/git\\"
  },
  \\"java\\": {
    \\"path\\": \\"/custom-path/java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"/custom-path/mvn\\"
  }
}
"
`;

exports[`setup initial prints an error if custom path is invalid 1`] = `"Failed to verify git executable at '/invalid/git/path', please ensure Git is properly installed. For more info see https://www.atlassian.com/git/tutorials/install-git"`;

exports[`setup initial prints an error if custom path is invalid 2`] = `"Failed to verify java executable at '/invalid/java/path'. Please ensure Java is properly installed. For more info see https://www.java.com/en/download/help/download_options.xml"`;

exports[`setup initial prints an error if custom path is invalid 3`] = `"Failed to verify mvn executable at '/invalid/maven/path'. Please ensure Maven is properly installed. For more info see https://maven.apache.org/install.html"`;
